framework/include/utils/SymmetricRankTwoTensor.h

```diff
@@ -0,0 +1,640 @@
+   * _vals[i][j] = R_ij * R_jl * _vals[k][l]
+   * @param R rotation matrix as a TypeTensor
+   */
+  void rotate(const TypeTensor<T> & R);
+
+  /**
+   * Returns a matrix that is the transpose of the matrix this
+   * was called on. This is a non-operation.
+   */
+  SymmetricRankTwoTensorTempl<T> transpose() const;
+
+  /// sets _vals to a, and returns _vals
+  SymmetricRankTwoTensorTempl<T> & operator=(const SymmetricRankTwoTensorTempl<T> & a);
+
+  /**
+   * Assignment-from-scalar operator.  Used only to zero out vectors.
+   */
+  template <typename Scalar>
+  typename boostcopy::enable_if_c<ScalarTraits<Scalar>::value, SymmetricRankTwoTensorTempl &>::type
+  operator=(const Scalar & libmesh_dbg_var(p))
+  {
+    libmesh_assert_equal_to(p, Scalar(0));
+    this->zero();
+    return *this;
+  }
+
+  /// adds a to _vals
+  SymmetricRankTwoTensorTempl<T> & operator+=(const SymmetricRankTwoTensorTempl<T> & a);
+
+  /// returns _vals + a
+  template <typename T2>
+  SymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+  operator+(const SymmetricRankTwoTensorTempl<T2> & a) const;
+
+  /// sets _vals -= a and returns vals
+  SymmetricRankTwoTensorTempl<T> & operator-=(const SymmetricRankTwoTensorTempl<T> & a);
+
+  /// returns _vals - a
+  template <typename T2>
+  SymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+  operator-(const SymmetricRankTwoTensorTempl<T2> & a) const;
+
+  /// returns -_vals
+  SymmetricRankTwoTensorTempl<T> operator-() const;
+
+  /// performs _vals *= a
+  SymmetricRankTwoTensorTempl<T> & operator*=(const T & a);
+
+  /// returns _vals*a
+  template <typename T2>
+  auto operator*(const T2 & a) const ->
+      typename std::enable_if<ScalarTraits<T2>::value,
+                              SymmetricRankTwoTensorTempl<decltype(T() * T2())>>::type;
+
+  /// performs _vals /= a
+  SymmetricRankTwoTensorTempl<T> & operator/=(const T & a);
+
+  /// returns _vals/a
+  template <typename T2>
+  auto operator/(const T2 & a) const ->
+      typename std::enable_if<ScalarTraits<T2>::value,
+                              SymmetricRankTwoTensorTempl<decltype(T() / T2())>>::type;
+
+  /// Defines multiplication with a vector to get a vector
+  template <typename T2>
+  TypeVector<typename CompareTypes<T, T2>::supertype> operator*(const TypeVector<T2> & a) const;
+
+  /// Defines logical equality with another SymmetricRankTwoTensorTempl<T2>
+  template <typename T2>
+  bool operator==(const SymmetricRankTwoTensorTempl<T2> & a) const;
+
+  /// Defines logical inequality with another SymmetricRankTwoTensorTempl<T2>
+  template <typename T2>
+  bool operator!=(const SymmetricRankTwoTensorTempl<T2> & a) const;
+
+  /// Sets _vals to the values in a ColumnMajorMatrix (must be 3x3)
+  SymmetricRankTwoTensorTempl<T> & operator=(const ColumnMajorMatrixTempl<T> & a);
+
+  /// returns _vals_ij * a_ij (sum on i, j)
+  T doubleContraction(const SymmetricRankTwoTensorTempl<T> & a) const;
```

Comment:
Probably for a future PR, also add `T doubleContraction(const RankTwoTensorTempl<T> & a) const;`

Reply:
Noted in #21269